ggml/src/ggml-vulkan/ggml-vulkan.cpp

```diff
@@ -4358,6 +4388,167 @@ static void ggml_vk_mul_mat(ggml_backend_vk_context * ctx, vk_context& subctx, c
     }
 }
 
+static void ggml_vkemb_mul_mat(ggml_backend_vk_context * ctx, vk_context& subctx, const ggml_tensor * src0, const ggml_tensor * src1, ggml_tensor * dst, bool dryrun = false) {
```

Comment:
I think this can probably be handled with the main `ggml_vk_mul_mat_q_f16` function. This would also give you the rest of the quants, since the function can do dequant in a separate shader + matmul in float16, and then you can avoid the separate backend and supports_op function.

Maybe switching the pipeline in `ggml_vk_guess_matmul_pipeline` and the parameter handling in `ggml_vk_matmul`?

Reply:
The problem is that we need to reject OPs for MUL_MAT with unsupported quants (the complex implementations doesn't work) and whole OPs like MUL_MAT_ID. Other OPs also have trouble with certain params. I think the requirement of having a separate `supports_op` function is going to be very hard to avoid.

> Maybe switching the pipeline in ggml_vk_guess_matmul_pipeline and the parameter handling in ggml_vk_matmul?

OK, let me see if I can make it fit.

Reply:
I guess if the matmul shaders are problematic, maybe the dequants also don't work, so some quants may just not work.

> The problem is that we need to reject OPs for MUL_MAT with unsupported quants (the complex implementations doesn't work) and whole OPs like MUL_MAT_ID. Other OPs also have trouble with certain params. I think the requirement of having a separate supports_op function is going to be very hard to avoid.

I understand. But at least you won't need to duplicate the backend if you add your function in front of the existing one and just jump directly to it if the device doesn't need the simple shaders.